src/librustc_lint/non_ascii_idents.rs

```diff
@@ -13,9 +14,58 @@ declare_lint! {
     "detects uncommon Unicode codepoints in identifiers"
 }
 
-declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS, UNCOMMON_CODEPOINTS]);
+declare_lint! {
+    pub CONFUSABLE_IDENTS,
+    Warn,
+    "detects uncommon Unicode codepoints in identifiers"
+}
+
+declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS, UNCOMMON_CODEPOINTS, CONFUSABLE_IDENTS]);
+
+fn calc_skeleton(symbol_str: &str) -> String {
+    if symbol_str == "ｓ" {
+        // FULLWIDTH LATIN SMALL LETTER S
+        return "s".to_owned();
+        // LATIN SMALL LETTER S
+    }
+    symbol_str.to_owned()
+}
 
 impl EarlyLintPass for NonAsciiIdents {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, _: &ast::Crate) {
+        use std::collections::BTreeMap;
+        let mut ascii_symbol_strs = BTreeMap::new();
+        let mut non_ascii_symbol_strs = BTreeMap::new();
+        let symbols = cx.sess.parse_sess.symbol_gallery.symbols.lock();
+        for symbol in symbols.iter() {
+            let symbol_str = symbol.as_str();
+            if symbol_str.is_ascii() {
+                ascii_symbol_strs.insert(symbol_str, *symbol);
+                continue;
```

Comment:
Nit: `if` with `else` would be more readable here.

Reply:
Actually, there's no need in these `BTreeMap`s (mapping `SymbolStr` to `Symbol`?!).
Everything can be collected into a single `Vec<SymbolStr>` and then `sort`ed before reporting.

Reply:
Yes, i think so, thanks.